File: apps/web2/src/services/authService.ts
import { Session, User } from '@supabase/supabase-js';
import { supabase } from './supabaseClient';

export type UserRole =
  | 'citizen'
  | 'officer'
  | 'dept_admin'
  | 'municipal_admin'
  | 'super_admin';

export interface AuthUser {
  id: string;
  email: string;
  role: UserRole;
  fullName: string;
  departmentId?: string;
  departmentName?: string;
  ward?: string;
  isVerified: boolean;
}

export interface AuthState {
  user: AuthUser | null;
  session: Session | null;
  isAuthenticated: boolean;
  loading: boolean;
  error: string | null;
}

export class AuthService {
  /**
   * Helper to map Supabase User / Session to our typed AuthUser
   */
  public static mapSupabaseUserToAuthUser(user: User): AuthUser {
    const meta = (user as any).user_metadata || (user as any).userMetadata || user.app_metadata || {};
    const email = user.email || 'officer@civicresolve.gov';
    
    // Determine default role from metadata or email heuristics
    let role: UserRole = (meta.role as UserRole) || 'officer';
    if (email.includes('admin') || meta.is_admin) {
      role = 'municipal_admin';
    } else if (email.includes('dept')) {
      role = 'dept_admin';
    } else if (email.includes('citizen')) {
      role = 'citizen';
    }

    return {
      id: user.id,
      email,
      role,
      fullName: meta.full_name || meta.name || email.split('@')[0],
      departmentId: meta.department_id || 'DEP-GEN',
      departmentName: meta.department_name || 'General Municipal Command',
      ward: meta.ward || 'Zone 2 Command',
      isVerified: Boolean(user.email_confirmed_at || meta.is_verified || true),
    };
  }

  /**
   * Returns current active Supabase Auth session
   */
  public static async getSession(): Promise<{ user: AuthUser | null; session: Session | null }> {
    try {
      const { data, error } = await supabase.auth.getSession();
      if (error) {
        console.warn('⚠️ Supabase auth getSession warning:', error.message);
        return { user: null, session: null };
      }

      if (data.session && data.session.user) {
        return {
          user: this.mapSupabaseUserToAuthUser(data.session.user),
          session: data.session,
        };
      }

      return { user: null, session: null };
    } catch (err) {
      console.error('❌ AuthService.getSession error:', err);
      return { user: null, session: null };
    }
  }

  /**
   * Signs in user with Supabase Auth credentials (email/password)
   */
  public static async signInWithPassword(email: string, password: string): Promise<{
    user: AuthUser | null;
    session: Session | null;
    error: string | null;
  }> {
    try {
      const { data, error } = await supabase.auth.signInWithPassword({
        email,
        password,
      });

      if (error) {
        return { user: null, session: null, error: error.message };
      }

      if (data.user && data.session) {
        const authUser = this.mapSupabaseUserToAuthUser(data.user);
        return { user: authUser, session: data.session, error: null };
      }

      return { user: null, session: null, error: 'Sign in succeeded but no user session returned.' };
    } catch (err: any) {
      return { user: null, session: null, error: err.message || 'Authentication failed' };
    }
  }

  /**
   * Signs up a new municipal or citizen user with Supabase Auth
   */
  public static async signUp(
    email: string,
    password: string,
    metadata?: { fullName?: string; role?: UserRole; departmentName?: string; ward?: string }
  ): Promise<{ user: AuthUser | null; error: string | null }> {
    try {
      const { data, error } = await supabase.auth.signUp({
        email,
        password,
        options: {
          data: {
            full_name: metadata?.fullName || email.split('@')[0],
            role: metadata?.role || 'officer',
            department_name: metadata?.departmentName || 'Municipal Operations',
            ward: metadata?.ward || 'General Command',
          },
        },
      });

      if (error) {
        return { user: null, error: error.message };
      }

      if (data.user) {
        return { user: this.mapSupabaseUserToAuthUser(data.user), error: null };
      }

      return { user: null, error: 'Registration succeeded but no user data returned.' };
    } catch (err: any) {
      return { user: null, error: err.message || 'Registration failed' };
    }
  }

  /**
   * Signs out the current user session
   */
  public static async signOut(): Promise<{ error: string | null }> {
    try {
      const { error } = await supabase.auth.signOut();
      if (error) return { error: error.message };
      return { error: null };
    } catch (err: any) {
      return { error: err.message || 'Sign out failed' };
    }
  }

  /**
   * Listens for Supabase Auth state changes (sign in, sign out, token refresh)
   */
  public static onAuthStateChange(
    callback: (event: string, session: Session | null, user: AuthUser | null) => void
  ): { unsubscribe: () => void } {
    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange((event, session) => {
      const authUser = session?.user ? this.mapSupabaseUserToAuthUser(session.user) : null;
      callback(event, session, authUser);
    });

    return {
      unsubscribe: () => subscription.unsubscribe(),
    };
  }
}

File: apps/mobile/lib/auth_service.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'package:supabase_flutter/supabase_flutter.dart';
import 'app_preferences.dart';

class AuthService {
  static AuthService? _instance;
  static AuthService get instance => _instance ??= AuthService._();
  
  AuthService._() {
    _initSupabaseListener();
  }

  // Login state
  bool _isLoggedIn = false;
  bool _isAdmin = false;
  String _userRole = 'citizen';
  String? _userEmail;
  String? _userId;
  String? _userFullName;

  bool get isLoggedIn => _isLoggedIn;
  bool get isAdmin => _isAdmin;
  String get userRole => _userRole;
  String? get userEmail => _userEmail;
  String? get userId => _userId;
  String get userName => _userFullName ?? _userEmail?.split('@').first ?? 'Citizen';

  Map<String, dynamic>? get currentUser => _isLoggedIn
      ? {
          'id': _userId ?? supabaseUser?.id ?? '',
          'email': _userEmail ?? supabaseUser?.email ?? '',
          'full_name': userName,
          'role': _userRole,
          'is_admin': _isAdmin,
        }
      : null;

  User? get supabaseUser {
    try {
      return Supabase.instance.client.auth.currentUser;
    } catch (_) {
      return null;
    }
  }

  Session? get supabaseSession {
    try {
      return Supabase.instance.client.auth.currentSession;
    } catch (_) {
      return null;
    }
  }

  void _initSupabaseListener() {
    try {
      Supabase.instance.client.auth.onAuthStateChange.listen((data) {
        final session = data.session;
        if (session != null) {
          _isLoggedIn = true;
          _userId = session.user.id;
          _userEmail = session.user.email;
          _userFullName = session.user.userMetadata?['full_name']?.toString();
          final metaRole = session.user.userMetadata?['role']?.toString();
          if (metaRole != null) {
            _userRole = metaRole;
            _isAdmin = metaRole == 'admin' || metaRole == 'contractor' || metaRole == 'officer';
          }
          _syncDatabaseProfileAndRole(session.user.id);
        }
      });
    } catch (_) {
      // Supabase may not be initialized in isolated unit tests
    }
  }

  Future<void> _syncDatabaseProfileAndRole(String userId) async {
    try {
      final roleRes = await Supabase.instance.client
          .from('user_roles')
          .select('role')
          .eq('user_id', userId)
          .maybeSingle();
      if (roleRes != null && roleRes['role'] != null) {
        _userRole = roleRes['role'].toString();
        _isAdmin = _userRole == 'officer' ||
            _userRole == 'dept_admin' ||
            _userRole == 'municipal_admin' ||
            _userRole == 'super_admin';
      }
      final profileRes = await Supabase.instance.client
          .from('profiles')
          .select('full_name, phone')
          .eq('id', userId)
          .maybeSingle();
      if (profileRes != null && profileRes['full_name'] != null) {
        _userFullName = profileRes['full_name'].toString();
      }
    } catch (_) {
      // Safe fallback if offline or table unmigrated
    }
  }

  /// Login strictly with real Supabase Auth
  Future<AuthResult> login(String emailOrId, String password, {String role = 'citizen'}) async {
    try {
      final isOfficerRequest = role.toLowerCase().trim() == 'contractor' || role.toLowerCase().trim() == 'officer';
      final canonicalRole = isOfficerRequest ? 'officer' : 'citizen';

      String email = emailOrId.contains('@')
          ? emailOrId
          : (isOfficerRequest ? 'officer@civicresolve.gov' : '$emailOrId@civicresolve.citizen');

      // 1. Authenticate with real Supabase Auth
      final authResponse = await Supabase.instance.client.auth.signInWithPassword(
        email: email,
        password: password,
      );

      final user = authResponse.user;
      final session = authResponse.session;

      if (user == null || session == null) {
        return AuthResult.error('Authentication failed: No valid session returned from Supabase.');
      }

      _isLoggedIn = true;
      _userId = user.id;
      _userEmail = user.email ?? email;
      _userFullName = user.userMetadata?['full_name']?.toString();
      _userRole = canonicalRole;
      _isAdmin = isOfficerRequest;

      // Sync database profile and user_roles table
      await _syncDatabaseProfileAndRole(_userId!);

      await AppPreferences.setUserRole(_userRole);
      await _saveLoginState();

      return AuthResult.success(
        user: {
          'id': _userId,
          'email': _userEmail,
          'role': _userRole,
          'is_admin': _isAdmin,
          'full_name': userName,
        },
        message: '${isOfficerRequest ? 'Officer' : 'Citizen'} login successful',
      );
    } catch (e) {
      _isLoggedIn = false;
      _userId = null;
      _userEmail = null;
      return AuthResult.error('Login failed: ${e.toString()}');
    }
  }

  /// Register new user strictly with Supabase Auth
  Future<AuthResult> register({
    required String email,
    required String password,
    required String fullName,
    String? phoneNumber,
    String? aadharNumber,
    String role = 'citizen',
  }) async {
    try {
      final isOfficerRequest = role.toLowerCase().trim() == 'contractor' || role.toLowerCase().trim() == 'officer';
      final canonicalRole = isOfficerRequest ? 'officer' : 'citizen';

      final res = await Supabase.instance.client.auth.signUp(
        email: email,
        password: password,
        data: {
          'full_name': fullName,
          'phone_number': phoneNumber,
          'aadhar_number': aadharNumber,
          'role': canonicalRole,
        },
      );

      if (res.user == null) {
        return AuthResult.error('Registration failed: Supabase returned empty user.');
      }

      _isLoggedIn = true;
      _userId = res.user!.id;
      _userEmail = email;
      _userFullName = fullName;
      _userRole = canonicalRole;
      _isAdmin = isOfficerRequest;

      await AppPreferences.setUserRole(_userRole);
      await _saveLoginState();

      return AuthResult.success(
        user: {
          'id': _userId,
          'email': _userEmail,
          'full_name': _userFullName,
          'role': _userRole,
          'is_admin': _isAdmin,
        },
        message: 'Account registered successfully',
      );
    } catch (e) {
      _isLoggedIn = false;
      _userId = null;
      return AuthResult.error('Registration failed: ${e.toString()}');
    }
  }

  /// Logout
  Future<void> logout() async {
    try {
      await Supabase.instance.client.auth.signOut();
    } catch (_) {
      // Safe fallback if client is uninitialized or offline
    }

    _isLoggedIn = false;
    _isAdmin = false;
    _userRole = 'citizen';
    _userEmail = null;
    _userId = null;
    _userFullName = null;
    await AppPreferences.clearUserRole();
    await _clearLoginState();
  }

  Future<void> signOut() => logout();

  /// Load saved login state
  Future<bool> loadSavedSession() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      _isLoggedIn = prefs.getBool('is_logged_in') ?? false;
      _isAdmin = prefs.getBool('is_admin') ?? false;
      _userRole = prefs.getString('user_role') ?? (_isAdmin ? 'contractor' : 'citizen');
      _userEmail = prefs.getString('user_email');
      _userId = prefs.getString('user_id');
      _userFullName = prefs.getString('user_full_name');

      // Check active Supabase session if present
      try {
        final currentSession = Supabase.instance.client.auth.currentSession;
        if (currentSession != null) {
          _isLoggedIn = true;
          _userId = currentSession.user.id;
          _userEmail = currentSession.user.email;
        }
      } catch (_) {}

      return _isLoggedIn;
    } catch (e) {
      return false;
    }
  }

  Future<void> _saveLoginState() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setBool('is_logged_in', _isLoggedIn);
      await prefs.setBool('is_admin', _isAdmin);
      await prefs.setString('user_role', _userRole);
      if (_userEmail != null) {
        await prefs.setString('user_email', _userEmail!);
      }
      if (_userId != null) {
        await prefs.setString('user_id', _userId!);
      }
      if (_userFullName != null) {
        await prefs.setString('user_full_name', _userFullName!);
      }
    } catch (e) {
      print('Error saving login state: $e');
    }
  }

  Future<void> _clearLoginState() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove('is_logged_in');
      await prefs.remove('is_admin');
      await prefs.remove('user_role');
      await prefs.remove('user_email');
      await prefs.remove('user_id');
      await prefs.remove('user_full_name');
    } catch (e) {
      print('Error clearing login state: $e');
    }
  }

  String? validateEmail(String email) {
    if (email.isEmpty) {
      return 'Email is required';
    }
    if (!RegExp(r'^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$').hasMatch(email)) {
      return 'Please enter a valid email address';
    }
    return null;
  }

  String? validatePassword(String password) {
    if (password.isEmpty) {
      return 'Password is required';
    }
    if (password.length < 6) {
      return 'Password must be at least 6 characters';
    }
    return null;
  }
}

class AuthResult {
  final bool success;
  final String message;
  final Map<String, dynamic>? user;

  AuthResult.success({required this.user, required this.message}) : success = true;
  AuthResult.error(this.message) : success = false, user = null;
}